Make topic and stream editing independent of `allow_message_editing` setting.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- FIrst commit is just to remove an unnecessary comment.
- Second and third commits are to add `Nobody` option to `edit_topic_policy` and `move_message_between_streams_policy` setting.
- Fourth commit is for API changes to make stream and topic editing independent of `allow_message_editing`.
- Fifth commit is for webapp changes to make stream and topic editing independent of `allow_message_editing`.
- Sixth commit is to update API to make topic editing permission independent of message sender and not allow message sender to edit topic indefinitely.
- Seventh commit is a refactor commit (explained in detail in commit message).
- Eigth commit is to update webapp to make topic editing permission independent of message sender and not allow message sender to edit topic indefinitely.

Fixes: <!-- Issue link, or clear description.--> Part of #21739.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented:
Need some changes in help documentaion, especially the table in https://zulip.com/help/configure-message-editing-and-deletion, but would probably be better to udpate it once we do remaining parts of #21739 especially-
> Controls all topic editing. The user's own messages will be treated exactly the same as anyone else's.

Also we need to update the documentation to reflect that we do not have "Never" option and instead we have a checkbox for `allow_message_editing` setting.

alya commented:
@timabbott I haven't reviewed, but I think you should just take a look.

alya commented:
Yeah, let's just update the documentation page once things are a bit more settled.

timabbott commented:
I did a few quick changes to comments and API docs and pushed back here. I'll post a few comments, some of which are likely a blocker for merging the later commits -- but in theory the first commit could be merged now I think?

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -61,6 +61,20 @@ def do_set_realm_property(
         value=value,
     )
+
+    message_edit_settings = [
```
Let's add a comment that these settings have a different event format due to their history.

sahil839 replied:
Added comment.

timabbott commented on zerver/lib/event_schema.py:
```diff
@@ -977,10 +977,20 @@ def check_realm_default_update(
+
+message_content_edit_limit_seconds_data = DictType(
+    required_keys=[
+        ("message_content_edit_limit_seconds", OptionalType(int)),
```
Why is this one `Optional`?

sahil839 replied:
We need this since the setting can be `None` for case when there is no limit. Previously it worked fine because we did not test `None` case, now also added a test for it.

timabbott replied:
OK great, thanks for confirming the previous state was a bug.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6204,6 +6204,11 @@ paths:
+        **Changes**: Before Zulip 6.0 (feature level 150), editing
+        streams and topics of messages was forbidden if
+        `allow_message_editing` was `false`, regardless of the
+        `edit_topic_policy` or `move_messages_between_streams_policy`.
```
So thinking about this change -- I believe it is removing entirely the ability to disable moving messages between streams/topics. I'm not sure we want to do that.

We could avoid that potential regression by first adding NOBODY to set the of options for `edit_topic_policy` and `move_message_between_streams_policy`. @alya thoughts?

alya replied:
Sure, yeah, that sounds like a good approach to me.

sahil839 replied:
Yes, can do that and would be easy change. But for topic editing case the user would still be allowed to edit topics sent by themselves (and also others using different `propogate_mode` values) since `edit_topic_policy` is only for message sent by others. We want to change this also though such that all messages are treated same and `edit_topic_policy` affects all messages as mentioned in the issue. So, I think I can also add the commit for that change too (along with adding `Nobody` option) and then we can move forward with this.

But probably we first need to complete #22652 and then would be simpler to refactor the code for the above changes.

timabbott commented:
Can you make sure we've had an API design conversation about the second change, to make the settings independent? We will at least want to coordinate with the mobile/terminal apps about whether they need to change their client logic (if it was previously correct) for showing users the edit options in the correct situations.

sahil839 commented:
Added comment as asked above and also answered couple of questions above.

Skimmed through mobile and terminal code, and probably terminal clients would need to change the logic but not mobile. Started discussion [here](https://chat.zulip.org/#narrow/stream/378-api-design/topic/make.20moving.20messages.20independent.20of.20allow_message_editing) for confirming.

sahil839 commented:
I think first commit can be merged if it looks good now. I have updated the feature level such that first commit can be merged independently.

timabbott commented:
Great, merged the first commit, which is the bulk of the code here, as 04693b6ac1b68be68570884ec881e38d83765be8. 

I think proceeding with #22652 as discussed https://github.com/zulip/zulip/pull/23056#discussion_r982667838 makes sense.

sahil839 commented:
This is still not ready for review. Need to add commit for making topic editing independent of message.

sahil839 commented:
@timabbott Updated the PR with following changes -
- Added commits to add `Nobody` option to `edit_topic_policy` and `move_messages_between_streams_policy` settings.
- Added commits to make topic editing permission independent of who sent the message.
- Also added a couple of refactor commits.

Some points/questions to note -
- As a result of second last commit, we now show `View source` button when neither content editing nor moving message is allowed even when `allow_message_editing` is `False` as discussed [here](https://chat.zulip.org/#narrow/stream/101-design/topic/message.20action.20shortcuts).
- Do we want to change the label of `Who can edit topic of any message` since permission do not depend on who sent the message now? Can also update the variables and functions in code similarly.

Now working on adding time limit setting and as discussed [here](https://chat.zulip.org/#narrow/stream/101-design/topic/time.20restrictions.20on.20topic.20editing/near/1448541). Not sure whether we want to merge this PR first without the time limit setting.

alya commented:
> Do we want to change the label of Who can edit topic of any message since permission do not depend on who sent the message now?

Yes, good thought!

How about this?
- "Who can edit topic of any message" -> "Who can move messages to another topic"
- "Who can move messages between streams" -> "Who can move messages to another stream"

This would be pretty consistent with how we describe these actions in the help center (https://zulip.com/help/move-content-to-another-topic , https://zulip.com/help/move-content-to-another-stream).

alya commented:
> Now working on adding time limit setting and as discussed [here](https://chat.zulip.org/#narrow/stream/101-design/topic/time.20restrictions.20on.20topic.20editing/near/1448541). Not sure whether we want to merge this PR first without the time limit setting.

I'm personally fine with that. Could we keep the current 3-day limit until the setting is implemented, or would that be awkward to do?

sahil839 commented:
Added two commits to update the label as well has function names.

This PR keeps the 3-day limit for topic editing for users who are not admins or moderators, there is no limit on changing streams which is same as previous behavior. The only major difference is that previously the message sender had permission to edit topics indefinitely but now message sender also has 3 day limit. Should be fine I guess.

alya commented:
Yeah, this seems like a reasonable checkpoint we can integrate to me.

alya commented:
Playing with the PR, "Enable message edit history" feels out of place now. How about if we move it just below "Allow message editing", above all the dropdowns?

alya commented:
I'm actually thinking "Message and topic editing" doesn't seem quite right anymore. Among other things, moving messages between streams is not really topic editing.

What if we had two separate sections?

## Message editing 
 Allow message editing
 Enable message edit history
Time limit for editing messages

## Moving messages
Who can move messages to another topic
Who can move messages to another stream
[when added] Time limit for moving messages

alya commented:
I'm not too worried about which help pages we link to for now for each of these sections, since https://zulip.com/help/configure-message-editing-and-deletion will need to be revamped anyhow.

sahil839 commented:
> I'm actually thinking "Message and topic editing" doesn't seem quite right anymore. Among other things, moving messages between streams is not really topic editing.
> 
> What if we had two separate sections?
> 
> ## Message editing
> Allow message editing Enable message edit history Time limit for editing messages
> 
> ## Moving messages
> Who can move messages to another topic Who can move messages to another stream [when added] Time limit for moving messages

Sounds good. Should be easy to change. Will add a commit to this PR soon.

sahil839 commented:
Added commit for adding separate sections. Here is the screenshot -
![Screenshot from 2022-10-18 23-10-26](https://user-images.githubusercontent.com/35494118/196509731-86ef6804-b48f-496c-8bf3-dbd5d262707c.png)

alya commented:
That screenshot looks good to me! @timabbott marking for your review.

alya commented:
Note that I looked at the settings panel in the PR, but did not try to manually verify that the permissions are working correctly, since this seems easier to do based on the code.

sahil839 commented:
Removing the "integration review" label for now. We should first merge #23307 to handle edge cases for locally echoed messages. Also moved some prep commits to that PR. Will update this PR after that and the label again, there should be now functional changes from the current state.

sahil839 commented:
Updated the PR. @timabbott ready for your review.

sahil839 commented:
Rebased the PR to solve merge conflicts.

timabbott commented on static/js/message_edit.js:
```diff
@@ -125,45 +125,54 @@ export function is_message_editable_ignoring_permissions(message) {
+export function get_editability(message, edit_limit_seconds_buffer = 0) {
+    if (!is_message_editable_ignoring_permissions(message)) {
+        return editability_types.NO;
+    }
```
I worry about `is_topic_editable` / `is_content_editable` not calling `is_message_editable_ignoring_permissions`; I feel like folks are going to call those functions directly and skip that check by accident. 

If it's really cheap, probably it'd be reasonable move the check into both functions? I wouldn't be surprised if that could be followed by a refactor that deletes `get_editability` in favor of calling those functions directly in most places -- the reason `get_editability` made sense as an interface was that there was a sort order to the possibilities.

timabbott commented:
This generally lgtm. I would like to figure out a plan for how to balance allowing users to fix typos in topics they themselves created with potential abuse issues involving moving topics within a stream following this change; we might at least want to change our documentation on moderating open communities with updated recommendations after integrating this. 

One potential rule that could help significantly is just "Any user can rename a topic that contains [at least as visible to themselves] only messages that they themselves sent for the first 4 hours after after sending."

sahil839 commented:
Added commit to call `is_message_editable_ignoring_permissions` in `is_content_editable` and also removed `get_editability` and instead use `is_content_editable`.

sahil839 commented:
> This generally lgtm. I would like to figure out a plan for how to balance allowing users to fix typos in topics they themselves created with potential abuse issues involving moving topics within a stream following this change; we might at least want to change our documentation on moderating open communities with updated recommendations after integrating this.
> 
> One potential rule that could help significantly is just "Any user can rename a topic that contains [at least as visible to themselves] only messages that they themselves sent for the first 4 hours after after sending."

We would need to do something to allow users to fix typos. I like the above suggested idea. Also, wanted to ask do we want to allow editing topics in such case even when topic editing is not allowed in any case by setting `edit_topic_policy` to nobody.

alya commented:
> One potential rule that could help significantly is just "Any user can rename a topic that contains [at least as visible to themselves] only messages that they themselves sent for the first 4 hours after after sending."

I don't understand the bracketed phrase. What does that mean?

timabbott commented:
OK, that can be a follow-up, but it should be a release goal and prioritized, since this change, which it makes the model much more sane, does significantly regress the typo correction experience for open communities, and anyone else who wants to limit which roles can edit topics.

timabbott commented:
Rebased because `main` had some CI failures and merged, thanks for all the work on this @sahil839!

alya commented:
Super excited to see major progress on this, by the way!

sahil839 commented:
> > OK, that can be a follow-up, but it should be a release goal and prioritized, since this change, which it makes the model much more sane, does significantly regress the typo correction experience for anyone who wants to limit which roles can
> > edit topics.
> 
> Looking back at #21739, our current plan is to have an independently customizable time limit for editing topics/moving messages, that would _only_ apply to members and below. @timabbott I think that solves precisely the problem you're raising here?

It will depend on whether we want to allow editing topics in such cases irrespective of other permissions and we want to have a separate time limit for case of only the user editing the topic has sent messages in that topic.

sahil839 commented:
> @sahil839 , with this PR merged, are we in a position to make the pencil icon in the message bar edit the topic for all messages, rather than just the ones below? Or do we need to make more changes for that not to be buggy?

Hmm, that should be easy to change with just changing the `propagate_mode` variable passed to the API. But I think would be better to first work on updating the work on moving messages which fall inside the time limit as [discussed here](https://chat.zulip.org/#narrow/stream/101-design/topic/time.20restrictions.20on.20topic.20editing). I have some work done for this and also for adding the new settings for time limits. Would probably be able to open PR for these in a couple of days once I am back from vacation next week.

timabbott commented:
Let's move this discussion to chat.zulip.org -- I think it deserves a larger audience. One thing we need to figure out is what precise semantics we want for time limits -- I think in an ideal world, if you're only going to be moving part of a topic due to a time limit, we want that to be a thing that's presented to you.

alya commented:
Chatted with @timabbott  offline -- we'll catch up/refresh the prior context, but a new discussion may not be necessary, since we have some clear next steps from last time we talked about this.

alya commented:
So @sahil839 I think you should not feel blocked on more discussion here.
